firmware/phy_task.c

```diff
@@ -0,0 +1,718 @@
+        {FW_PHY_EVENT_PHYMSTR_REQ, NULL, NULL, NULL , &dfiPhyMstrPending},
+        {FW_PHY_EVENT_PHYUPD_REQ, NULL, NULL, NULL, &dfiPhyUpd},
+        {FW_PHY_EVENT_CTRLUPD_ASSERT, NULL, NULL, NULL, &dfiCtrlUpd},
+    },
+    .numTransitions = 3,
+    .data = NULL,
+    .entryAction = NULL,
+    .exitAction = NULL,
+};
+
+static struct state dfiPhyMstrPending = {
+    .parentState = NULL,
+    .entryState = NULL,
+    .transitions = (struct transition[]) {
+        {FW_PHY_EVENT_PHYMSTR_ACK, NULL, NULL, NULL, &dfiPhyMstr},
+    },
+    .numTransitions = 1,
+    .data = NULL,
+    .entryAction = dfi_phymstr_pending_entry_handler,
+    .exitAction = NULL,
+};
+
+static struct state dfiPhyMstr = {
+    .parentState = NULL,
+    .entryState = NULL,
+    .transitions = (struct transition[]) {
+        {FW_PHY_EVENT_PHYMSTR_EXIT, NULL, NULL, NULL, &dfiIdle},
+    },
+    .numTransitions = 1,
+    .data = NULL,
+    .entryAction = dfi_phymstr_enter_handler,
+    .exitAction = dfi_phymstr_exit_handler,
+};
+
+static struct state dfiPhyUpd = {
+    .parentState = NULL,
+    .entryState = NULL,
+    .transitions = (struct transition[]) {
+        {FW_PHY_EVENT_PHYUPD_ACK, NULL, NULL, dfi_phyupd_update, &dfiIdle},
+    },
+    .numTransitions = 1,
+    .data = NULL,
+    .entryAction = dfi_phyupd_entry_handler,
+    .exitAction = NULL,
+};
+
+static struct state dfiCtrlUpd = {
+    .parentState = NULL,
+    .entryState = NULL,
+    .transitions = (struct transition[]) {
+        {FW_PHY_EVENT_CTRLUPD_DEASSERT, NULL, NULL, NULL, &dfiIdle},
+    },
+    .numTransitions = 1,
+    .data = NULL,
+    .entryAction = dfi_ctrlupd_entry_handler,
+    .exitAction = NULL,
+};
+
+/*******************************************************************************
+**                              IMPLEMENTATIONS
+*******************************************************************************/
+void fw_phy_task_init(void)
+{
+    configASSERT(fw_manager.mq == NULL);
+    configASSERT(fw_manager.task == NULL);
+
+    // Early initialization of PHY
+    phy_early_init();
+
+    // Initialize device and drivers
+    wddr_init(&wddr, WDDR_BASE_ADDR, &table);
+
+    // State machine initialization
+    stateM_init(&fw_manager.fsm.fsw, &fswPrep, &errorState);
+    stateM_init(&fw_manager.fsm.dfi, &dfiIdle, &errorState);
+
+    // Create message queue
+    fw_manager.mq = xQueueCreate(MSG_QUEUE_LEN, sizeof(fw_msg_t));
+    configASSERT(fw_manager.mq != NULL);
+
+    // Create PHYUPD Timer
+    xDfiPhyUpdTimer = xTimerCreate("DFI PHYUPD Timer",
+                                    DFI_PHYUPD_PERIOD,
+                                    pdTRUE,
+                                    NULL,
+                                    dfi_phyupd_timer_callback);
+    configASSERT(xDfiPhyUpdTimer != NULL);
+
+    // Create the task
+    xTaskCreate(firmwareTask,
+                "FW Task",
+                configMINIMAL_STACK_SIZE,
+                NULL,
+                configMAX_PRIORITIES - 1,
```

Comment:
Because of the abstraction, I don't want the priority to be exposed at the app level. however, we do need to figure out our story with priorities. We can revisit this later.